geo/geogfn: implement ST_Project

hueypark opened the pull request:
Fixes #48402

Release note (sql change): This PR implement adds the ST_Project({geography,float8,float8})

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49949)
<!-- Reviewable:end -->

otan commented on pkg/geo/geogfn/unary_operators.go:
```diff
@@ -90,6 +93,41 @@ func Length(g *geo.Geography, useSphereOrSpheroid UseSphereOrSpheroid) (float64,
+	}
+
+	// Normalize azimuth
+	azimuth -= 2.0 * math.Pi * math.Floor(azimuth/(2.0*math.Pi))
```
if we are using `s1.Angle()`, there is a `Normalized` function we can use.

hueypark replied:
Done.

otan commented on pkg/geo/geogfn/unary_operators.go:
```diff
@@ -90,6 +93,41 @@ func Length(g *geo.Geography, useSphereOrSpheroid UseSphereOrSpheroid) (float64,
+
+	// Check the distance validity.
+	if distance > (math.Pi * spheroid.Radius) {
+		return nil, errors.Newf("Distance must not be greater than %v", math.Pi*spheroid.Radius)
```
nit: errors start with lowercase.
does this error also appear in PostGIS? can we use %f instead?'

hueypark replied:
Done.
Yes, PostGIS has the same error.
https://github.com/postgis/postgis/blob/65095c6c0eaa188cb2b36a6274d6a9c77cf7c5ea/liblwgeom/lwgeodetic.c#L2119

otan commented on pkg/geo/geogfn/unary_operators.go:
```diff
@@ -90,6 +93,41 @@ func Length(g *geo.Geography, useSphereOrSpheroid UseSphereOrSpheroid) (float64,
+	y := point.Y()
+
+	projected := spheroid.Project(
+		s2.LatLng{Lat: s1.Angle(x), Lng: s1.Angle(y)},
```
i believe you can use s2.LatLngFromDegrees(point.Y(), point.X()). I also Lat is x, Lng is y.

hueypark replied:
Done.

otan commented on pkg/geo/geogfn/unary_operators.go:
```diff
@@ -128,3 +166,65 @@ func length(regions []s2.Region, useSphereOrSpheroid UseSphereOrSpheroid) (float
 }
+
+// longitudeNormalize convert a longitude to the range of -Pi, Pi.
+func longitudeNormalize(lon float64) float64 {
```
can we convert this to `s1.Angle`, and use `s1.Angle(blah).Normalized()`?

hueypark replied:
Done.

otan commented on pkg/geo/geographiclib/geographiclib.go:
```diff
@@ -109,3 +114,24 @@ func (s *Spheroid) AreaAndPerimeter(points []s2.Point) (area float64, perimeter
+
+	C.geod_direct(
+		&s.cRepr,
+		C.double(math.Pi*point.Lat/180.0),
```
we can use `point.Lat.Degrees()` for this, i believe.

hueypark replied:
Done.

otan commented on pkg/geo/geographiclib/geographiclib.go:
```diff
@@ -109,3 +114,24 @@ func (s *Spheroid) AreaAndPerimeter(points []s2.Point) (area float64, perimeter
+		nil)
+
+	return s2.LatLng{
+		Lat: s1.Angle(float64(lat) * math.Pi / 180.0),
```
s1.Angle(lat * s1.Degree) may be good here.

hueypark replied:
I fixed this with s2.LatLngFromDegrees.

otan commented on pkg/sql/sem/builtins/geo_builtins.go:
```diff
@@ -854,6 +854,55 @@ var geoBuiltins = map[string]builtinDefinition{
+The distance is given in meters. Negative values are supported.
+
+The azimuth (also known as heading or bearing) is given in radians. It is measured clockwise from true north (azimuth zero).
+East is azimuth π/2 (90 degrees); south is azimuth π (180 degrees); west is azimuth 3π/2 (270 degrees).
```
if you wanted extra new lines here to display in HTML format, you'll need an extra new line here.
This new line is otherwise displayed as spaces (see autogenerated code and how the <p> tags are missing.)

hueypark replied:
I've seen `functions.md`, but I can't find anything wrong about the format. Can you explain it in more detail?

otan replied:
```

<p>The azimuth (also known as heading or bearing) is given in radians. It is measured clockwise from true north (azimuth zero).
East is azimuth π/2 (90 degrees); south is azimuth π (180 degrees); west is azimuth 3π/2 (270 degrees).
Negative azimuth values and values greater than 2π (360 degrees) are supported.</p>
```

if you wanted new line between them, you'd want

```

<p>The azimuth (also known as heading or bearing) is given in radians. It is measured clockwise from true north (azimuth zero).</p>
<p>East is azimuth π/2 (90 degrees); south is azimuth π (180 degrees); west is azimuth 3π/2 (270 degrees).</p>
<p>Negative azimuth values and values greater than 2π (360 degrees) are supported.</p>
```

then you need a new line between each here. if you're ok with just space difference, then this is ok. i'm up for either.

hueypark replied:
Thank you for explain. I want to keep it as it is.

otan reviewed (changes requested):
Thanks! I have a few suggestions on making this cleaner.

hueypark commented:
Thank you for the review.

otan commented on pkg/geo/geogfn/unary_operators.go:
```diff
@@ -128,3 +166,32 @@ func length(regions []s2.Region, useSphereOrSpheroid UseSphereOrSpheroid) (float
 }
+
+// latitudeNormalize convert a latitude to the range of -Pi/2, Pi/2.
+func latitudeNormalize(lat float64) float64 {
```
nit: can you name this `normalizeLatitude`?
also shame there isn't something similar in s1/s2 :(

hueypark replied:
Done.

otan commented on pkg/geo/geographiclib/geographiclib.go:
```diff
@@ -109,3 +114,22 @@ func (s *Spheroid) AreaAndPerimeter(points []s2.Point) (area float64, perimeter
+		C.double(distance),
+		&lat,
+		&lng,
+		nil)
```
nit: i prefer trailing commas and one arg per line, i.e.

```
  nil,
)
```

easier to git blame and less merge conflict-y.

hueypark replied:
Done.

otan reviewed (commented):
two small things -- otherwise LGTM! thank you so much!

otan commented on pkg/geo/geogfn/unary_operators.go:
```diff
@@ -90,6 +93,41 @@ func Length(g *geo.Geography, useSphereOrSpheroid UseSphereOrSpheroid) (float64,
+	projected := spheroid.Project(
+		s2.LatLngFromDegrees(x, y),
+		distance,
+		azimuth)
```
trailing comma here too, and on geom.NewPointFlat

hueypark replied:
Done.

hueypark committed:
geo/geogfn: implement ST_Project

Fixes #48402

Release note (sql change): This PR implement adds the ST_Project({geography,float8,float8})

otan commented:
bors r+
